### PR TITLE
Northstar client refactoring

### LINKED ISSUE #77: Niche: incorrect mapping to new Northstar user
Niche importer passes various fields incorrectly to Northstar. For example:
- Field [`address1`](https://github.com/DoSomething/mbc-user-import/blob/master/src/MBC_UserImport_Source_Niche.php#L172) should be mapped `addr_street1`
- Field [`address2`](https://github.com/DoSomething/mbc-user-import/blob/master/src/MBC_UserImport_Source_Niche.php#L175) should be mapped `addr_street2`
- Field [`city`](https://github.com/DoSomething/mbc-user-import/blob/master/src/MBC_UserImport_Source_Niche.php#L178) should be mapped `addr_city`
- Field [`state`](https://github.com/DoSomething/mbc-user-import/blob/master/src/MBC_UserImport_Source_Niche.php#L181) should be mapped `addr_state`
- Field [`zip`](https://github.com/DoSomething/mbc-user-import/blob/master/src/MBC_UserImport_Source_Niche.php#L189) should be mapped `addr_zip`

Because of that this information is not saved.
Northstar documentation: https://github.com/DoSomething/northstar/blob/dev/documentation/endpoints/users.md#create-a-user

cc @mshmsh5000 @justkika @jamjensen

### PR DESCRIPTION
**Pull request by sergii-tkachenko**
#### What's this PR do?
- Refactors Niche to use Northstar library
- Improves Niche user import flow

#### Where should the reviewer start?
👁 

#### What are the relevant tickets?
Fixes #77, #93, #96.
Trello: https://trello.com/c/ZIQmj1sT/208-8-niche-refactor-to-northstar-gateway-library

### CODE REVIEW

File: src/MBC_UserImport_Source_Niche.php [hunk 1]
```diff
@@ -254,209 +195,270 @@ public function setter($message)
...
-    // Check for existing email account in MailChimp
-    $subscribed = $this->mbcUserImportToolbox->checkExistingEmail(
-      $this->importUser,
-      $existing
-    );
-    if (!$subscribed) {
-      $this->addEmailSubscriptionSettings($this->importUser, $payload);
+    // Shortcuts.
+    $northstar = &$this->northstar;
+    $input = &$this->user;
```
Comment (DFurnes) [inline src/MBC_UserImport_Source_Niche.php]: This feels like it's unnecessarily obscuring things for the sake of saving a few keystrokes.

File: src/MBC_UserImport_Source_Niche.php [hunk 2]
```diff
@@ -254,209 +195,270 @@ public function setter($message)
...
-      $northstarUser = $this->mbToolbox->createNorthstarUser($importUser);
+      // Create Northstar and Phoenix accounts.
+      self::log(
+        'User not found, creating new user on Northstar: %s',
+        json_encode($input)
+      );
+      $identity = $northstar->createUser($input);
+    } elseif (!empty($identityByEmail) && empty($identityByMobile)) {
+      // ****** Existing user: only email record exists ******
+      $identity = &$identityByEmail;
```
Comment (DFurnes) [inline src/MBC_UserImport_Source_Niche.php]: Same here… is there an advantage to using a reference?

File: src/MBC_UserImport_Source_Niche.php [hunk 3]
```diff
@@ -254,209 +195,270 @@ public function setter($message)
...
 
-    // Campaign signup
-    $campaignNID = self::PHOENIX_SIGNUP;
-    $campaignSignup = $this->mbcUserImportToolbox->campaignSignup(
-      $campaignNID,
-      $drupalUID,
-      'niche',
-      false
-    );
+    } elseif ($identityByEmail->id === $identityByMobile->id) {
```
Comment (DFurnes) [inline src/MBC_UserImport_Source_Niche.php]: Four of these cases ("neither email nor mobile exist", "email exists w/o mobile, add new unclaimed mobile", "mobile exists w/o email, add new unclaimed email" and "email and mobile already exist on same account") should be handled by Northstar's `POST v1/users` upsert logic: if it finds any of those three cases, it'll create or update the existing record with the new fields.

For the case where two separate users already exist and you need to choose which to use, Northstar will return a `422` validation error and it'd make sense to only manually resolve the conflict when that occurs.

Comment (DFurnes) [inline src/MBC_UserImport_Source_Niche.php]: And if relying on the upsert functionality on that endpoint, you can then just use the HTTP status code to determine whether the account was newly created by this import: 200 for previously existed, 201 for new.

File: src/MBC_UserImport_Source_Niche.php [hunk 4]
```diff
@@ -254,209 +195,270 @@ public function setter($message)
...
-   * @param object $drupalUser The user object created by Drupal API.
-   *
-   * @return null
-   */
-  public function addImportUserInfo($drupalUser)
-  {
-
-    $this->importUser['uid'] = $drupalUser->drupal_id;
-  }
+  /** Bad OOP is bad OOP */
```
Comment (DFurnes) [inline src/MBC_UserImport_Source_Niche.php]: 😂

File: src/MBC_UserImport_Source_Niche.php [hunk 5]
```diff
@@ -109,140 +120,70 @@ public function canProcess($message)
...
-    }
-    if (!empty($message['password'])) {
-      $this->importUser['merge_vars']['PASSWORD'] = $message['password'];
+      $this->user['last_name'] = $message['last_name'];
     }
+
+    // Address.
     if (!empty($message['address1'])) {
-      $this->importUser['address1'] = $message['address1'];
+      $this->user['addr_street1'] = $message['address1'];
```
Comment (sergii-tkachenko) [inline src/MBC_UserImport_Source_Niche.php]: Yes, good idea. Usually, I do it like this, just too used to legacy Quicksilver stile.

File: src/MBC_UserImport_Source_Niche.php (hunk 2 above)
Comment (sergii-tkachenko) [inline src/MBC_UserImport_Source_Niche.php]: `$identity` was not a shortcut, but a `let`-kind-of variable. Passing it by reference: yes, this was not necessary, just a micro-optimization.

File: src/MBC_UserImport_Source_Niche.php (hunk 3 above)
Comment (sergii-tkachenko) [inline src/MBC_UserImport_Source_Niche.php]: We can simplify it later, for now I'd like to keep it here for the logging.

Comment (sergii-tkachenko) [inline src/MBC_UserImport_Source_Niche.php]: I created an issue for it: https://github.com/DoSomething/mbc-user-import/issues/99

Comment (DFurnes) [inline src/MBC_UserImport_Source_Niche.php]: Sounds good! 👍

**Pull request merged**